tests/scripts/icmp_responder.py

```diff
@@ -1,72 +1,139 @@
-from scapy.data import ETH_P_IP
-from select import select
-
-
-def respond_to_icmp_request(socket, request, dst_mac=None):
-    """Respond to ICMP request."""
-    reply = request.copy()
-    reply[ICMP].type = 0
-    # Force re-generation of the checksum
-    reply[ICMP].chksum = None
-    reply[IP].src, reply[IP].dst = request[IP].dst, request[IP].src
-    reply[IP].chksum = None
-
-    if dst_mac is None:
-        icmp_reply_dst_mac = request[Ether].src
-    else:
-        icmp_reply_dst_mac = dst_mac
-
-    reply[Ether].src, reply[Ether].dst = request[Ether].dst, icmp_reply_dst_mac
-    socket.send(reply)
-
-
-class ICMPSniffer(object):
-    """Sniff ICMP packets."""
-
-    TYPE_ECHO_REQUEST = 8
-    def __init__(self, iface, request_handler=None, dst_mac=None):
-        """
-        Init ICMP sniffer.
-
-        @param ifaces: interfaces to listen for ICMP reqest
-        @param request_handler: handler function that will be called when
-                                receives ICMP request
-        """
-        self.sniff_sockets = []
-        self.iface_hwaddr = {}
-        self.sniff_sockets.append(conf.L2socket(type=ETH_P_IP, iface=iface, filter="icmp"))
-        self.iface_hwaddr[iface] = get_if_hwaddr(iface)
-        self.request_handler = request_handler
+from scapy.all import Ether, ICMP, IP
+
+
+# As defined in asm/socket.h  
+SO_ATTACH_FILTER = 26 
+
+# BPF filter "icmp[0] == 8"
+icmp_bpf_filter = [
+    [ 0x28, 0, 0, 0x0000000c ],
+    [ 0x15, 0, 8, 0x00000800 ],
+    [ 0x30, 0, 0, 0x00000017 ],
+    [ 0x15, 0, 6, 0x00000001 ],
+    [ 0x28, 0, 0, 0x00000014 ],
+    [ 0x45, 4, 0, 0x00001fff ],
+    [ 0xb1, 0, 0, 0x0000000e ],
+    [ 0x50, 0, 0, 0x0000000e ],
+    [ 0x15, 0, 1, 0x00000008 ],
+    [ 0x6, 0, 0, 0x00040000 ],
+    [ 0x6, 0, 0, 0x00000000 ]
+]
+
+
+def bpf_stmt(code, jt, jf, k):
+    """Format struct `sock_filter`."""
+    return struct.pack("HBBI", code, jt, jf, k)
+
+
+def build_bpfilter(filter):
+    """Build BPF filter buffer."""
+    return ctypes.create_string_buffer(b"".join(bpf_stmt(*_) for _ in filter))
+
+
+class ICMPResponderProtocol(asyncio.Protocol):
+    """ICMP responder protocol class to define read/write callbacks."""
+
+    def __init__(self, on_con_lost, dst_mac=None):
+        self.transport = None
+        self.on_con_lost = on_con_lost
         self.dst_mac = dst_mac
 
-    def __call__(self):
-        try:
-            while True:
-                sel = select(self.sniff_sockets, [], [])
-                for s in sel[0]:
-                    packet = s.recv()
-                    if packet is not None:
-                        if ICMP in packet and packet[ICMP].type == self.TYPE_ECHO_REQUEST and self.request_handler:
-                            self.request_handler(s, packet, self.dst_mac)
-        finally:
-            for s in self.sniff_sockets:
-                s.close()
+    def connection_made(self, transport):
+        self.transport = transport
 
+    def data_received(self, data):
```

Comment:
Where is `data_received` called?

Reply:
It is a callback hook, so if a protocol class is used to bind to a socket, its overloaded hooks will be called if any read/write event is polled from the socket, pls refer to: https://docs.python.org/3/library/asyncio-protocol.html.